google/cloud/forseti/notifier/notifiers/gcs_violations.py

```diff
@@ -29,30 +30,46 @@
 class GcsViolations(base_notification.BaseNotification):
     """Upload violations to GCS."""
 
-    def _get_output_filename(self):
-        """Create the output filename.
+    def _upload_json(self, gcs_upload_path):
+        """Upload violations in json format.
 
-        Returns:
-            str: The output filename for the violations CSV file.
+        Args:
+            gcs_upload_path (string): the GCS upload path.
         """
-        now_utc = date_time.get_utc_now_datetime()
-        output_timestamp = now_utc.strftime(
-            string_formats.TIMESTAMP_TIMEZONE_FILES)
-        output_filename = string_formats.VIOLATION_CSV_FMT.format(
-            self.resource, self.cycle_timestamp, output_timestamp)
-        return output_filename
+        with tempfile.NamedTemporaryFile() as tmp_violations:
+            tmp_violations.write(parser.json_stringify(self.violations))
+            tmp_violations.flush()
+            storage_client = storage.StorageClient()
+            storage_client.put_text_file(tmp_violations.name, gcs_upload_path)
 
-    def run(self):
-        """Generate the temporary CSV file and upload to GCS."""
+    def _upload_csv(self, gcs_upload_path):
+        """Upload violations in csv format.
+
+        Args:
+            gcs_upload_path (string): the GCS upload path.
+        """
         with csv_writer.write_csv(resource_name='violations',
                                   data=self.violations,
                                   write_header=True) as csv_file:
             LOGGER.info('CSV filename: %s', csv_file.name)
+            storage_client = storage.StorageClient()
+            storage_client.put_text_file(csv_file.name, gcs_upload_path)
 
-            gcs_upload_path = '{}/{}'.format(
-                self.notification_config['gcs_path'],
-                self._get_output_filename())
-
-            if gcs_upload_path.startswith('gs://'):
-                storage_client = storage.StorageClient()
-                storage_client.put_text_file(csv_file.name, gcs_upload_path)
+    def run(self):
+        """Generate the temporary (CSV xor JSON) file and upload to GCS."""
+        if self.notification_config['gcs_path'].startswith('gs://'):
+            data_format = self.notification_config.get('data_format', 'csv')
+            if data_format not in ['json', 'csv']:
```

Comment:
This is the second time this is checked, see my comment above in `_make_content()`.  I think it's better to check and raise here early on.

Reply:
Done.